wordpress-mobile/WordPress-Android
# Configure Codecov

## Description

This PR integrates Codecov. By this, we'll have:
- comments on each PR reporting code coverage diff (e.g. https://github.com/woocommerce/woocommerce-android/pull/10523#issuecomment-1879634587)
- a trend metrics available at the dashboard: https://app.codecov.io/github/wordpress-mobile/WordPress-Android/

![Screenshot 2024-01-12 at 16 39 56](https://github.com/wordpress-mobile/WordPress-Android/assets/5845095/62c371d6-fc57-4534-8ed0-55617b2cef1c)


## To Test:

1. Open https://app.codecov.io/github/wordpress-mobile/WordPress-Android/
2. See that for this PR, there's a reprt and it contains some code coverage data for some class from each module:
    - `WordPress`
    - `:libs:image-editor`
    - `:libs:processors`

It's double-check. From what I've browsed, all seems fine. Also, each job on Buildkite reports successful report upload to Codecov.

## Decission log

### Why specify the report file directly?

Codecov couldn't find reports if they were named differently than `report.xml`. So initially, it worked only for a JVM module (`:lib:processors`) but didn't work for e.g. `WordPress`, because the report file is named `reportWordpressVanillaRelease.xml`.

### Why use `CODECOV_TOKEN` if WordPress-Android is a public repository?

It's possible to use Codecov uploader without specifing a token if repository is open source. This approach is flaky though and results with often upload failures. Details are available in this comment and thread: https://github.com/codecov/codecov-action/issues/557#issuecomment-1216749652
